lambda/tmc.ml

```diff
@@ -0,0 +1,854 @@
+        (* non-recursive bindings are not specialized *)
+        let def = traverse ctx def in
+        let+ body = choice ctx ~tail body in
+        Llet (lk, vk, var, def, body)
+    | Lletrec (bindings, body) ->
+        let ctx, bindings = traverse_letrec ctx bindings in
+        let+ body = choice ctx ~tail body in
+        Lletrec(bindings, body)
+    | Lswitch (l1, sw, loc) ->
+        (* decompose *)
+        let consts_lhs, consts_rhs = List.split sw.sw_consts in
+        let blocks_lhs, blocks_rhs = List.split sw.sw_blocks in
+        (* transform *)
+        let l1 = traverse ctx l1 in
+        let+ consts_rhs = choice_list ctx ~tail consts_rhs
+        and+ blocks_rhs = choice_list ctx ~tail blocks_rhs
+        and+ sw_failaction = choice_option ctx ~tail sw.sw_failaction in
+        (* rebuild *)
+        let sw_consts = List.combine consts_lhs consts_rhs in
+        let sw_blocks = List.combine blocks_lhs blocks_rhs in
+        let sw = { sw with sw_consts; sw_blocks; sw_failaction; } in
+        Lswitch (l1, sw, loc)
+    | Lstringswitch (l1, cases, fail, loc) ->
+        (* decompose *)
+        let cases_lhs, cases_rhs = List.split cases in
+        (* transform *)
+        let l1 = traverse ctx l1 in
+        let+ cases_rhs = choice_list ctx ~tail cases_rhs
+        and+ fail = choice_option ctx ~tail fail in
+        (* rebuild *)
+        let cases = List.combine cases_lhs cases_rhs in
+        Lstringswitch (l1, cases, fail, loc)
+    | Lstaticraise (id, ls) ->
+        let ls = traverse_list ctx ls in
+        Choice.return (Lstaticraise (id, ls))
+    | Ltrywith (l1, id, l2) ->
+        (* in [try l1 with id -> l2], the term [l1] is
+           not in tail-call position (after it returns
+           we need to remove the exception handler),
+           so it is not transformed here *)
+        let l1 = traverse ctx l1 in
+        let+ l2 = choice ctx ~tail l2 in
+        Ltrywith (l1, id, l2)
+    | Lstaticcatch (l1, ids, l2) ->
+        (* In [static-catch l1 with ids -> l2],
+           the term [l1] is in fact in tail-position *)
+        let+ l1 = choice ctx ~tail l1
+        and+ l2 = choice ctx ~tail l2 in
+        Lstaticcatch (l1, ids, l2)
+    | Levent (lam, lev) ->
+        let+ lam = choice ctx ~tail lam in
+        Levent (lam, lev)
+    | Lifused (x, lam) ->
+        let+ lam = choice ctx ~tail lam in
+        Lifused (x, lam)
+
+  and choice_apply ctx ~tail apply =
+    let exception No_tmc in
+    try
+      match apply.ap_func with
+      | Lvar f ->
+          let explicit_tailcall_request =
+            match apply.ap_tailcall with
+            | Default_tailcall -> false
+            | Tailcall_expectation true -> true
+            | Tailcall_expectation false ->
+                (* [@tailcall false] disables TMC optimization
+                   on this tailcall *)
+                raise No_tmc
+          in
+          let specialized =
+            try Ident.Map.find f ctx.specialized
```

Comment:
The arity of specialized should be checked to be equal to the length of the list of arguments

Reply:
```OCaml
  type t = T : (int -> t) -> t

  let[@tail_mod_cons] rec map u n =
    print_int n;
    print_newline ();
    T (map u)

  let t = map () 3

  let T f = t

  let _ = f 4
```